flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java

```diff
@@ -601,41 +602,44 @@ public LocatableInputSplitAssigner getInputSplitAssigner(FileInputSplit[] splits
 		if (unsplittable) {
 			int splitNum = 0;
 			for (final FileStatus file : files) {
+				String bomCharsetName = getBomCharset(file);
```

Comment:
> I'm not sure if we want to check the BOM during split generation.
> 
> 1. This might become a bottleneck, since splits are generated by the JobManager. OTOH, there is currently an effort to parallelize split generation.
> 2. FileInputFormat is currently not handling any charset issues.
> 
> An alternative would be to check the BOM in `DelimitedInputFormat` when a split is opened.

@fhueske Hi, fhueske, if you check the BOM in DelimitedInputFormat when opening the split, I think the following should be considered:
1. A file is split into different TaskManagers, then the BOM of the verification file is required on each TaskManager.

Reply:
Yes, I'm aware of that. It would also be required for every split unless we cache the BOM per file.
OTOH, if we do it in the JM, the job cannot start until a single thread had a look at the first bytes of each file.

Reply:
Well, I first take the logic of checking the BOM in FileInputFormat to DelimitedInputFormat. I want to use a Map to cache the BOM encoding of the file, using the filename as the key and the BOM encoding as the value. If the value exists in the Map, the corresponding value is read, and if the Map does not exist, the BOM encoding of the file is read.

Reply:
Yes, I think that makes sense.

Reply:
Whether to judge the encoding type of a file or byte stream (no BOM), there may be the following scenario:
The file type is encoded as UTF-8, but the user-specified encoding type is UTF-16 or UTF-32, which still causes incorrect garbled parsing. On the other hand, it is a nuisance to determine what type of encoding a file or byte stream (no BOM) is.

Reply:
I don't think we should do too much magic here. If the specified encoding does not match the actual encoding of a file, we could reject it (either logging the error and continuing or failing) since we cannot infer a different encoding with 100% certainty.

Reply:
@fhueske I have developed the code logic. Do I want to resubmit the PR or reopen the PR?

Reply:
@fhueske I resubmitted a PR with the following link address:
[https://github.com/apache/flink/pull/6823](https://github.com/apache/flink/pull/6823)
Thank you very much, I hope you can help me review this PR again. Use your precious time.